### ISSUE TITLE
Set number of threads after numpy import

### ISSUE THREAD

**Issue by paugier** [numpy/numpy#11826]
This is not a bug report but just an enhancement proposal.

I think it would be useful and important to be able to easily set the number of threads used by Numpy after Numpy import.

From the perspective of library developers, it is often useful to be able to control the number of threads used by Numpy, see for example https://github.com/biopython/biopython/pull/1401.

It is not difficult to do in a simple script when we are sure that Numpy or Scipy have not been imported previously with something like:

```python
import os
os.environ["OMP_NUM_THREADS"] = "1"
import numpy as np
```

However, in a library there is a good chance that the user has already imported Numpy in its main script, with something like

```python
import numpy as np

import fluidimage
```

In this case, I don't see how to set the number of threads used by Numpy from the fluidimage code.

Thus, it would be very convenient to have a function `np.set_num_threads`.

**Comment by mattip** [comment]
We would have to make this pluggable somehow to adapt for different linalg implementations. For instance, OpenBLAS seems to expose a [`openblas_set_num_threads`](https://stackoverflow.com/questions/29559338/set-max-number-of-threads-at-runtime-on-numpy-openblas), but MKL would use some other function

**Comment by mattip** [comment]
@touqir14 see the [developer documentation](http://www.numpy.org/devdocs/dev/index.html). You should write some [tests](http://www.numpy.org/devdocs/reference/testing.html), tests that try out the new functions, verifying that they indeed set the number of threads you desire (perhaps by writing a C-level tests-only function that calls `NPY_BEGIN_THREADS` and ends with `NPY_END_THREADS`, and in the middle somehow counts the number of threads started) and then write code to implement the functions.

**Comment by juliantaylor** [comment]
note that this is actually quite tricky to tackle.
NumPy does not actually know which BLAS library is used to implement the functions, it just assumes that whatever implements it uses standard BLAS apis and abis. As there unfortunately is no standard api to determine information on the provider numpy would be relient on runtime introspection to determine the actual provider of the functions. As there are not very many of them (some I recall are openblas, atlas, blis, mkl and reference blas) it might be possible but still difficult to get working portably.

**Comment by touqir14** [comment]
Thanks for the heads up!

**Comment by FabianIsensee** [comment]
This is an extremely annoying problem because I want to explicitly turn off multiprocessing for my background workers and as far as I can see there is no way of properly doing that (except downgrading numpy to < 1.14). OMP_NUM_THREADS=1 will solve my problem for the most part, but it will prevent my entire python process, including the main thread, from using openmp multiprocessing while I would like to only disable it in the background workers. As of now, I have 8 background workers (multiprocessing.Process), each of which will spawn another 8 threads to do np.dot computation (which is only a very small part of what they are actually doing). That clogs up the entire CPU. Any suggestions of how I could solve this?

**Comment by mattip** [comment]
@FabianIsensee The problem is that NumPy does not know right now what linalg backend you are using, OpenBlas, MKL, ATLAS, ... Did you see this stackoverflow question mentioned [above](https://github.com/numpy/numpy/issues/11826#issuecomment-416579660), which suggests functions you can wrap and call to control this for your backend.

**Comment by FabianIsensee** [comment]
@mattip Thanks for pointing that out! That is a good solution to the problem, provided that the system you are running on uses openBLAS. Unfortunately this solution has no effect on my system. What's especially annoying for me is that I am providing an open source framework that runs into this problem and I cannot know what blas library each and every one of the users is going to use.
I am not familiar with the numpy releases but the multi-threaded matmul appear starting from version 1.14.6. I checked the release notes and could not spot anything that would have enabled that (or am I blind?). Since multithreading these operations did not happen prior to that there must be a way to control that from within numpy (none of my libraries changed)

**Comment by charris** [comment]
The change in 1.14.6 was building manylinux1 against OpenBLAS instead of ATLAS, we were already using OpenBLAS for Windows. As far as the build goes, 1.14.6 is pretty much 1.15.x. It sounds like we need to find a solution to this that doesn't depend on the user knowing which library is in use, probably adding some function in numpy that stores its info during the build.

**Comment by charris** [comment]
> Would something like this be possible?

I don't know, but it sounds like we need it. If nothing else we can pass info upstream and try to encourage the libraries themselves to add such a feature, ideally as some sort of standard library interface in BLAS (LAPACK) itself. @njsmith IIRC, you reported that there was some work going on to produce a new standard? In any case, @matthew-brett I think we could have some effect on OpenBLAS. Maybe there is already such a feature.

**Comment by charris** [comment]
Although with a single dynamic library I don't see how one could coordinate between callers. Hmm, not a simple problem, almost something that needs to be handled at the OS level. This is getting beyond my expertise.

**Comment by mattip** [comment]
@FabianIsensee this seems troubling
>Unfortunately this solution has no effect on my system
What exactly did you try and what was the result?

**Comment by FabianIsensee** [comment]
@mattip I did exactly what was decribed in the comment you referenced above.
```
    with num_threads(1):
        import numpy as np
        a = np.random.random((20000000, 3))
        b = np.random.random((3, 3))
        for _ in range(10):
            c = np.dot(a, b)
```

```
    import numpy as np
    a = np.random.random((20000000, 3))
    b = np.random.random((3, 3))
    for _ in range(10):
        c = np.dot(a, b)
```

These are the two examples I compared. I ran them while looking at htop to see CPU usage. For both of them a number of threads was spawned and CPU usage was above 100% for the main thread.

Running it like this `OMP_NUM_THREADS=1 python playground/numpy_threading_problem.py` gives the intended behavior on my computer, but this may not transfer to other machines.

**Comment by stuarteberg** [comment]
Whether or not a numpy API is feasible for this feature, perhaps we can crowdsource a new section in the numpy docs to explain this issue and offer advice with respect to the different environment variables.

IIUC, the basic advice for OpenBLAS and MKL users would be: If you're planning to use multiple processes, set `OMP_NUM_THREADS=1` before running your program.  If not, you can leave it unset.  (It would be nice to explain why this is important, too.)

Beyond that, maybe an MKL expert can offer more fine-grained advice with respect to these variables:
https://software.intel.com/en-us/mkl-linux-developer-guide-intel-mkl-specific-environment-variables-for-openmp-threading-control

For ATLAS users, apparently the number of threads is predetermined at compile time:
http://math-atlas.sourceforge.net/faq.html#tnum

@charris wrote:
>Although with a single dynamic library I don't see how one could coordinate between callers.

I don't quite grok this point, which is all the more reason I would love to see some docs on this general topic.  In which section of the numpy docs should I start a PR on the topic of multithreading control?

**Comment by bbbbbbbbba** [comment]
> IIUC, the basic advice for OpenBLAS and MKL users would be: If you're planning to use multiple processes (e.g. via `multiprocessing`, `dask`, `pyspark`, etc.), set `OMP_NUM_THREADS=1` before running your program. If not, you can leave it unset. (It would be nice to explain why this is important, too.)

Correct me if I'm wrong, but it's not a big problem with _multiprocessing_, right? At worst you pay some extra overhead for parallelizing more than you have CPU cores, but each process spawns its own threads and work correctly. It's _multithreading_ that is really problematic and causes bugs like #11046.

**Comment by FabianIsensee** [comment]
I would very much disagree here. In the specific situation that I am in, I have a pool of background workers that generate batches for a deep learning algorithm. These batches contain images where some of these images need to be (among other things) rotated for data augmentation (which is implemented via matrix multiplication). The machine I am working on is a dgx1 computer with 8 graphics cards and 80 
 CPU threads. Usually I train 8 different networks on them with 10 workers and 1 GPU each. Now each of these workers (80 in total) will attempt to do these matrix multiplications (which are quite tiny by the way) in a multithreaded way and since each worker sees 80 cpus they will spawn 80 threads each, resulting in the system being completely clogged up. That effectively breaks everything for me and the only way I can continue my word is to downgrade to numpy 1.14.5

**Comment by stuarteberg** [comment]
>Correct me if I'm wrong, but it's not a big problem with multiprocessing, right? At worst you pay some extra overhead for parallelizing more than you have CPU cores

I think we're on the same page, @bbbbbbbbba.  In the case of multiprocessing, I'm not worried about incorrect results.  As you said, I'm worried about poor performance due to spawning more threads than you can schedule onto your CPU cores.

But the penalty is not trivial!  In one of my recent use-cases on a 16-core machine, the performance was 6x worse due to the extra threads.  On an 80-core machine like @FabianIsensee's, the overhead must be even worse.

The threading-related issue you referred to is troubling, but that sounds like an outright bug, not an issue with the `OMP_NUM_THREADS` setting.  Hopefully someone will eventually implement a fix in `OpenBLAS` or a workaround `numpy`.

**Comment by mattip** [comment]
[This code](https://github.com/tomMoral/loky/pull/135/files#diff-e49a2eb30dd7db1ee9023b8f7306b9deR110) which is MIT licensed and based on other BSD-licensed routines at runtime probes the loaded DLLs (shared objects) to find which implementation is relevant and calls the implementation specific routine to set the number of threads. Thanks to @ogrisel for [this comment](https://github.com/xianyi/OpenBLAS/issues/1844#issuecomment-434955556) pointing it out

**Comment by ogrisel** [comment]
> This code which is MIT licensed and based on other BSD-licensed routines at runtime probes the loaded DLLs (shared objects) to find which implementation is relevant and calls the implementation specific routine to set the number of threads. Thanks to @ogrisel for this comment pointing it out

Sorry for being late to the party I had not seen this issue. Indeed we started to investigate over-subscription issues a bit in the context of scikit-learn / joblib but this is still work in progress. Having numpy expose a uniform API to control the behavior of the underlying BLAS thread pool would be nice. Note that @anton-malakhov and @tomMoral are the primary authors of those dynamic ctypes-based access to the underlying runtime libraries.

Ping @jeremiedb who might also be interested in following this discussion.

**Comment by seberg** [comment]
Wondering if such a snipplet/approach as above is completely nonsense or not? But at least maybe it can help someone to easier answer the "what BLAS are you using" question, if someone knows details for other libs feel free to edit post. Accelerate could be interesting?

**Comment by mattip** [comment]
@seberg the script I linked to uses C-types and OS-provided functions to  walk down the loaded shared objects looking for the one we want. Isn't that what your script does, only using Popen([ldd ...)`? It seems that if the clib provides [dl_iterate_phdr]https://linux.die.net/man/3/dl_iterate_phdr) for linux, [_dyld_image_count](http://www.manpagez.com/man/3/dyld/) for MacOS and [GetModuleFileNameExW](https://docs.microsoft.com/en-us/windows/desktop/api/psapi/nf-psapi-getmodulefilenameexa) for windows we should use them.

**Comment by seberg** [comment]
Yeah, was wondering a bit if it makes sense to try to avoid walking the dlls. One reason was to be specific to numpy, but I am sure that is solvable. The other was that the code used the module name only and at least for the numpy I got currently installed, that name is just blas/cblas and not openblas, so some (symbol) inspection would still be necessary?

That said, I guess the module name often will normally be sufficient, and I do not know linking and C-library, etc. details enough to really figure out what can or cannot work reliably.

**Comment by seberg** [comment]
Ah sorry, forget about the ldd stuff, I just added it because I kept looking at it. No, the first function just loads the multiarray.so with ctypes and checks if certain function symbols are defined... That seems to work for openblas, mkl, blis and atlas. But I have no idea if just trying to load function symbols should work, or if e.g. accelerate is identifiable by the existance of such a symbol.

**Comment by Amir-Arsalan** [comment]
So isn't it still possible to set the number of threads after importing `numpy`?

**Comment by seberg** [comment]
@Amir-Arsalan that depends on the BLAS that is being used, but in most cases, yes it is possible. Using ctypes it shouldn't be hard, especially if you know which BLAS you are using. Putting an as reliable as possible utility function into numpy has its tricky parts (plus for ATLAS and maybe Accelerate it just is not possible to adjust threads).

But, most users end up using either OpenBLAS or MKL (via Anaconda) and if you do not aim for perfect reliability I think creating a function that works for your system is very straight forward.

**Comment by Amir-Arsalan** [comment]
@seberg So I guess the safest way is still to set the environmental variables. Is that right?

**Comment by seberg** [comment]
@Amir-Arsalan, depends on what you need probably. The code Matti linked should work on most systems (I think it probably fails on debian). Except windows, my stuff might even work pretty reliably. I don't think the problem is getting pretty far. So if the environment variables have issues or you want to be able to switch the behaviour, its worth to do this kind of stuff. Just don't expect it to work on all setups (but neither do the environment variables I guess). Frankly, I don't know shared library linking/loading well enough to really understand how much can or cannot go wrong.

**Comment by Amir-Arsalan** [comment]
@seberg I just need to enable Numpy to use multiple threads/cores for its operations. One example would be doing matrix inversion. By environmental variables, I meant these:

```
export MKL_NUM_THREADS=8
export NUMEXPR_NUM_THREADS=8
export OMP_NUM_THREADS=8
```

**Comment by tomMoral** [comment]
@Amir-Arsalan The environment variable are the most reliable and cross platform way to control the number of threads for given implementations and if you want to set it for a given interpreter. You can use:
```bash
OMP_NUM_THREADS: openmp,
OPENBLAS_NUM_THREADS: openblas,
MKL_NUM_THREADS: mkl,
VECLIB_MAXIMUM_THREADS: accelerate,
NUMEXPR_NUM_THREADS: numexpr
```
It is possible to set all these variables to control the number of threads, without knowing which library is loaded in advance.

The issue is when you want dynamic control over the number of threads used. Indeed, once the thread pool is started, changing the environment variable will have no effect. Then, you need to access C-library functionality exposed to set the number of threads used, if they are provided by the underlying implementation. @seberg codes works on linux system, where `find_library` is able to locate and load the correct library file (`.so`) to call the dynamic thread-pool rescaling functions. The code in [a PR for `loky`](https://github.com/tomMoral/loky/pull/135/files#diff-e49a2eb30dd7db1ee9023b8f7306b9deR110) try to do it in a cross platform way, by looping through all DLL and setting the number of threads when it is possible. But both implementations have shortcomings, notably they will fail for some platforms (debian?) and implementations (`Accelerate` has no way to set the number of threads dynamically).

I am currently working on improving the `loky` code to make it the most reliable possible but it probably needs feedback. I am open to any suggestions for improvements or better support and comments if you see big drawbacks in the proposed way of changing the number of threads used.

**Comment by seberg** [comment]
@tomMoral, on my Debian and Arch just probing if the thread setting symbols exist in the numpy `.so` file(s) seems to work fine. On windows it fails. But I have no idea how reliable it can be assumed to work even on linux (The later code I had linked in the now hidden comments seems to distinguish blis, openblas, mkl, and atlas ok on linux/debian).

**Comment by mimakaev** [comment]
We're facing this problem when developing a library that uses multiprocessing.Pool and numpy.   We couldn't find any reliable way to set number of threads to 1. `mkl.set_num_threads(1) `  does not work if mkl is not installed. While setting environmental variables is best for your own machine, it is not really an option for a library: `os.environ["OMP_NUM_THREADS"] = "1"` has to be done before importing numpy. So if a user imported numpy and then imported our library, we're screwed. Finally, ctypes solution requires knowing library name, which is different in windows/linux/MacOS and for different versions of optimizations. Solution for Ubuntu/MKL is `ctypes.CDLL("libmkl_rt.so").mkl_set_num_threads(ctypes.byref(ctypes.c_int(1)))`  -  it works after numpy was imported, making it safer for a library.  We are likely to go with all 3 for now hoping one of them works.

**Comment by seberg** [comment]
@mimakaev I don't know it well, but I think the code that I gave (basically loading the numpy SO instead of "libmkl_rt.so") probably works even more reliable on linux, I think I list the function names in openblas and blis to set the threads, but if not I can look it up. I don't know linking lookup well, so I am a bit hesitant, but it could be cool to just gather a collection of hacks if that achieves near 100% correctness.

**Comment by FabianIsensee** [comment]
Hi,
what is the status of this issue? I would very much like to use newer numpy versions but this is a complete blocker as of right now! If some package makes use of multithreading then there must be a way to control the number of threads for the user. numpy really messes up CPU utilization on high CPU count servers!
Best,
Fabian

**Comment by seberg** [comment]
@FabianIsensee numpy does not use multithreading on its own for exactly this reason. However, lthere are two exceptions:
 1. Linear algebra/OpenBLAS (which numpy uses internally) can, and also the wheels will do this. You can use `OMP_NUM_THREADS=1` to disable that. Depending on the BLAS you are using, other steps may be necessary (e.g. not sure this is sufficient for MKL).
 2. There are Intel patches in NumPy linked against MKL, which also make it into Anaconda. In that case some numpy internal functions (outside linear algebra) may be patched to use multiple threads. I expect `OMP_NUM_THREADS` may work, there is likely even a monkeypatched additional functions to change it run-time, but this is outside of the normal numpy implementation, so I am not willing to check right now.

If you use anaconda, and the Intel patches are an issue, quite frankly, you may want to open an Anaconda issue or ping them here, so that they know that the idea of using threaded stuff everywhere is not trivially a good idea.

**Comment by anton-malakhov** [comment]
@FabianIsensee I think this is exactly what my threading composability talks are about. Please refer to https://github.com/IntelPython/composability_bench/tree/master/scipy2018_demo
In short, we've implemented two approaches as 'smp' and 'tbb4py' modules, which provide command line interface in order to modify application behavior w.r.t. number of threads or task scheduler used underneath. joblib folks develop a good alternative 'loky', which might want to use the same approach as smp module: https://github.com/tomMoral/loky/pull/135

Intel's patches also include mklinit module which solves co-existence and initialization problems of OpenMP runtime from different vendors.

It's better to open issues directly at https://github.com/IntelPython/numpy

**Comment by 1313e** [comment]
> what is the status of this issue? I would very much like to use newer numpy versions but this is a complete blocker as of right now! If some package makes use of multithreading then there must be a way to control the number of threads for the user. numpy really messes up CPU utilization on high CPU count servers!

Simply execute `export OMP_NUM_THREADS=1` before running your Python script and you solved the problem.
I had the same problem about a year ago, where adding more processes would result in slower code execution due to more OpenMP threads being requested on a single core.

However, I would like to see a way to manually control the number of threads that NumPy's linalg module uses, as that would still result in a significant speed-up for my code.

**Comment by FabianIsensee** [comment]
Hi everyone,
please excuse my very late reply and thank you very much for your responses! I will look into @anton-malakhov 's repository, hopefully it does what I need.
@1313e , @seberg I am aware of the `OMP_NUM_THREADS` 'fix', but what I really need is a solution along the lines of 'numpy.set_omp_threads(1)'. I am providing an open source python package (batchgenerators) that uses multiprocessing.Process to parallelize data augmentation for deep neural networks (that's usually 8 or so workers. If this runs on an 40C/80T machine then each worker will spawn 80 threads for linear algebra such as rotating a coordinate mesh via matrix multiplication = up to 640 threads. That clogs up everything). I cannot expect our users to set `OMP_NUM_THREADS` whenever they are using our package. I am basically in the same situation as I believe @mimakaev is in :-/
Best,
Fabian

**Comment by anton-malakhov** [comment]
there is no API for you unfortunately, the tools were designed as command-line helpers for users. loky might be a better shot if you need an API. We discussed with sklearn folks that when loky will implement functionality from smp module, it can be rebased on top of loky. Overall, there is a need for an industry-wide KB, parallelism orchestration library, similar to what Apache Arrow does for memory layout of data frames, otherwise users will be swamped under env vars like MKL_NUM_THREADS, NUMPY_NUM_THREADS, NUMEXPR_NUM_THREADS.. and so on. API approach is good only if you know with whom you are synchronizing and if you can restore the defaults (otherwise there is a risk of affecting other modules), so it's rather context manager than a potential `numpy.set_num_threads()`

**Comment by 1313e** [comment]
@FabianIsensee I know that feeling, as I am in the same situation.
The way I fixed that is that I check if `OMP_NUM_THREADS` was set to 1 whenever my package is imported and it uses more than a single MPI process, and if not, it raises a warning telling you that you might want to set it to 1.
Of course, it is not a perfect solution, as the user will have to set it to 1 first and then re-execute their script, but at least you remind them of what they have to do.

**Comment by mimakaev** [comment]
We will probably use this code as a solution for now. It is hard to test this because you never know how are people/distributions compiling numpy. But this code I believe should work at least for MKL.  

```
import os, ctypes 
def set_mkl_threads():
   
    try:
        import mkl
        mkl.set_num_threads(1)
        return 0
    except:
        pass 
    
    for name in [ "libmkl_rt.so", "libmkl_rt.dylib", "mkl_Rt.dll"]: 
        try: 
            mkl_rt = ctypes.CDLL(name)
            mkl_rt.mkl_set_num_threads(ctypes.byref(ctypes.c_int(1)))
            return 0 
        except:
            pass   

    os.environ["OMP_NUM_THREADS"] = "1" # export OMP_NUM_THREADS=4
    os.environ["OPENBLAS_NUM_THREADS"] = "1" # export OPENBLAS_NUM_THREADS=4 
    os.environ["MKL_NUM_THREADS"] = "1" # export MKL_NUM_THREADS=6
    os.environ["VECLIB_MAXIMUM_THREADS"] = "1" # export VECLIB_MAXIMUM_THREADS=4
    os.environ["NUMEXPR_NUM_THREADS"] = "1" # export NUMEXPR_NUM_THREADS=6

set_mkl_threads()
```

There is also a nearly-complete solution somewhere in the link below, but we didn't use it in the end because it seemed complicated. 
https://github.com/tomMoral/loky/pull/135/files#diff-e49a2eb30dd7db1ee9023b8f7306b9deR110

**Comment by tomMoral** [comment]
I am writing a dedicated package joblib/threadpoolctl with the solution we wrote in loky (tommoral/loky#135). It currently handles `mkl`, `openmp` and `openblas` with both a context API and a functional one. I did not finish packaging it to pypi and it is still in heavy development but if some of you want to test it from master and give me some feedback, it would be awesome.

**Comment by paugier** [comment]
@tomMoral `threadpoolctl` would be great to solve my original issue without Numpy modification! Thank you for this great work! We'll use it in [fluidimage](https://bitbucket.org/fluiddyn/fluidimage) when it's ready (we should actually also use joblib/loky for executors of topologies but it's another story).

I tried `threadpoolctl` and it seems to work nicely with the setup I used to try it (Debian, pyenv, numpy/scipy installed with pip). However, I observed something strange so I filled an issue in your bug tracker.
